File: pwauth.py
#!/usr/bin/python

import crypt
import sys

db = {
  'users': [
  {
    'id': 1,
    'user': 'user1',
    'domain_id': 1,
    'password': 'testpw1',
    'hpw': 'a9cXlAGYp1YTg'
  },
  {
    'id': 2,
    'user': 'user2',
    'domain_id': 1,
    'password': 'testpw2',
    'hpw': 'E.cTWtAAuY4dg'
  }
  ],
  'domains': [
  {
    'id': 1,
    'domain': 'domain1.com'
  },
  {
    'id': 2,
    'domain': 'domain2.com'
  }
  ]
}

def usage():
    sys.stderr.write( sys.argv[0] + ' <username> <password>\n' )
    sys.exit(0)

def authUser( username, pw ):
    at = username.find( '@' )
    if at < 1:
      return 'Invalid user/password1 for ' + username

    user = username[:at]
    domain = username[at+1:]

    user_obj = getUser( user, domain )
    if user_obj is None:
      return 'Invalid user/password2 for ' + username

    if checkPassword( user_obj, pw ) != 1:
      return 'Invalid user/password3 for ' + username

    print 'Authentication successful for %s' % ( username )
    print 'User: %s\nDomain: %s\nUsername: %s' % ( user_obj['user'], user_obj['domain'], username )
    return None

def checkPassword( user, password ):
    if crypt.crypt( password, user['hpw'] ) == user['hpw']:
      return True

    return False

def isValidDomain( domain ):
    domain_obj = None

    for key in db['domains']:
      if domain == key.get('domain',None):
        domain_obj = key
        break
    
    return domain_obj

def getUser( user, domain ):
    dKey = isValidDomain( domain )

    if dKey is None:
      return None

    userid = None
    for key in db['users']:
      if user == key.get('user',None):
        userid = key
        break

    if userid is None:
      return None

    user_obj = dict(
      dKey.items() + userid.items()
    )

    return user_obj


if len(sys.argv) < 3:
    usage()

error = authUser( sys.argv[1], sys.argv[2] )
if error is not None:
    sys.stderr.write( error + '\n' )
    sys.exit(1)

sys.exit(0)

File: pwauth.pl
#!/usr/bin/perl
#

use strict;
use Data::Dumper;

my %db = (
  'domains' => {
    '1' => { 'domain' => 'domain1.com' },
    '2' => { 'domain' => 'domain2.com' }
  },
  'users' => {
    '1' => {
      'user' => 'user1',
      'domain' => 1,
      'password' => 'testpw1',
      'hpw' => 'a9cXlAGYp1YTg'
    },
    '2' => {
      'user' => 'user2',
      'domain' => 1,
      'password' => 'testpw2',
      'hpw' => 'E.cTWtAAuY4dg'
    }
  }
);

usage() if( @ARGV < 2 );

my( $user, $passwd ) = @ARGV;

authUser( $user, $passwd );

exit 0;

sub usage {
  print STDERR "$0 <username> <password>\n";
  exit 1;
}

sub authUser {
  my( $username, $pw ) = @_;

  my $at = index( $username, '@' );
  die "Invalid user/password1 for $username\n" if( $at < 0 || $at > length( $username ) );

  my $user = substr( $username, 0, $at );
  my $domain = substr( $username, $at + 1 );

  my $user_obj = getUser( $user, $domain );
  die "Invalid user/password2 for $username\n" unless( $user_obj );

  die "Invalid user/password3 for $username\n" unless( checkPassword( $user_obj, $pw ) );

  print "Authentication successful for $username\n";
  printf "User: %s\nDomain: %s\nUsername: %s\n",
    $user_obj->{'user'},
    $user_obj->{'domain'},
    $username;
}

sub checkPassword {
  my( $user, $password ) = @_;

  return 1 if( crypt( $password, $user->{'hpw'} ) eq $user->{'hpw'} );
  return 0;
}

sub isValidDomain {
  my( $domain ) = @_;

  foreach my $domain_key ( keys %{$db{'domains'}} ) {
    return $domain_key if( $db{'domains'}{$domain_key}{'domain'} eq $domain );
  }

  return undef;
}

sub getUser {
  my( $user, $domain ) = @_;

  my $dKey = isValidDomain( $domain );
  return undef unless( $dKey );

  my $userid = undef;
  foreach my $user_key ( keys %{$db{'users'}} ) {
    next unless( $db{'users'}{$user_key}{'user'} eq $user );

    $userid = $user_key;
    last;
  }

  return undef unless( defined $userid );

  my %user_obj = (
    %{$db{'users'}{$userid}},
    %{$db{'domains'}{$db{'users'}{$userid}{'domain'}}}
  );
  return \%user_obj;
}
